src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
+    end
+    y = first(last(hobbits))
+    axe = x isa AbstractArray ? axes(x) : nothing
+    type = Val(typeof(x))
+    project = x isa AbstractArray{<:Number} ? ProjectTo(x) : identity
+    function unfoldl(dy)
+        _tini = (nothing, unthunk(dy), nothing)
+        trio = accumulate(_reverse1(hobbits); init=_tini) do (_,dc,_), (_,back)
+            ds, da, db = back(dc)  # 's' for self, don't need to write LHS
+            # Don't need to store every `da`, need one for the next iteration + maybe last
+        end
+        df = sum(first, trio)
+        dx = map(t -> t[3], _reverse1(trio))
+        if init === _InitialValue()
+            # `hobbits` is one short
+            dx = _vcat1(trio[end][2], dx)
+        end
+        dx_nice = _val(type) <: Tuple ? Tangent{_val(type)}(dx...) : reshape(dx, axe)
+        return (NoTangent(), df, project(dx_nice))
+    end
+    return y, unfoldl
+end
+
+if VERSION >= v"1.6"
+    _reverse1(x) = Iterators.reverse(x)
+    _drop1(x) = Iterators.drop(x, 1)
+    _zip2(x, y) = zip(x, y)  # for `accumulate`, below
+else
+    # Old versions don't support accumulate(::itr), nor multi-dim reverse
+    _reverse1(x) = reverse(vec(x))
+    _drop1(x) = vec(x)[1:end-1]
+    _zip2(x, y) = collect(zip(x, y))
+end
+_reverse1(x::Tuple) = reverse(x)
+_drop1(x::Tuple) = Base.tail(x)
+_zip2(x::Tuple{Vararg{Any,N}}, y::Tuple{Vararg{Any,N}}) where N = ntuple(i -> (x[i],y[i]), N)
+
+struct _InitialValue end  # Old versions don't have Base._InitialValue
+
+_vcat1(x, ys::AbstractVector) = vcat(x, ys)
+_vcat1(x::AbstractArray, ys::AbstractVector) = vcat([x], ys)
+_vcat1(x, ys::Tuple) = (x, ys...)
+
+_no_tuple_tangent(dx::Tangent) = ChainRulesCore.backing(dx)
+_no_tuple_tangent(dx) = dx
+
+
+#####
+##### `accumulate`
+#####
+
+# Like `foldl` this by definition works in order, so it makes sense to allow stateful `f`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(accumulate), f::F, x::Union{AbstractArray, Tuple}; 
+        init=_InitialValue(), dims=nothing
+    ) where {F}
+    isnothing(dims) || dims==1 && x isa Base.AbstractVecOrTuple || throw(
+        "accumulate(f, x; dims) is not supported by ChainRules, sorry"
+        # It's not supported by AD either, so no point calling back, and no regression:
+        # gradient(x -> sum(accumulate(/, x, dims=1)), rand(3,4)) # ERROR: Mutating arrays is not supported
+    )
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init = (start, nothing)) do (a,_), b
+        c, back = rrule_via_ad(config, f, a, b)
+    end
+    y = map(first, hobbits)
+    if init === _InitialValue()
+        # `hobbits` is one short, and first one doesn't invoke `f`
+        y = _vcat1(first(x), y)
+    end
+    axe = x isa AbstractArray ? axes(x) : nothing
+    type = Val(typeof(x))
+    project = x isa AbstractArray{<:Number} ? ProjectTo(x) : identity
```

Comment:
per above, if we made a helper functor for carrying this information and applying it, we could reuse it here for cleanlyness here also